engine/classes/Elgg/Collections/Collection.php

```diff
@@ -0,0 +1,270 @@
+<?php
+
+namespace Elgg\Collections;
+
+use ArrayAccess;
+use Countable;
+use InvalidParameterException;
+use SeekableIterator;
+
+/**
+ * A collection of unique items
+ */
+class Collection implements CollectionInterface,
+							ArrayAccess,
+							SeekableIterator,
+							Countable {
+
+	/**
+	 * @var CollectionItemInterface[]
+	 */
+	protected $items = [];
+
+	/**
+	 * @var string
+	 */
+	protected $item_class;
+
+	/**
+	 * Constructor
+	 *
+	 * @param CollectionItemInterface[] $items      Items
+	 * @param string                    $item_class Member class
+	 *                                              Restrict members of the collection to instances of this class
+	 */
+	public function __construct($items = [], $item_class = CollectionItemInterface::class) {
+		$this->item_class = $item_class;
+
+		foreach ($items as $item) {
+			$this->add($item);
+		}
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function all() {
+		return $this->items;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function count() {
+		return count($this->items);
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function add($item) {
+		if (!$item instanceof $this->item_class) {
+			elgg_log("Collection can only contain instances of " . $this->item_class, 'ERROR');
+			return $this;
+		}
+
+		$this->items[$item->getId()] = $item;

```

Comment:
If this collection is constructed with `\ElggObject` as the item class this will fail.

Eighter add a check here that $item implements `CollectionItemInterface` or make the constructor through an exception if `$item_class` doesn't implement `CollectionItemInterface`

Reply:
I was throwing an exception but then PHPStorm keeps highlight stuff with Unhandled exception. But yes, you have a point.